File: windows/src/ControlDeck.Domain/Ids.cs
namespace ControlDeck.Domain;

/// <summary>
/// Strongly-typed identifier wrappers. All ControlDeck ids are UUID strings on
/// the wire (see protocol/PROTOCOL.md §2); wrapping them prevents accidentally
/// passing a DeviceId where a WidgetId is expected, etc.
/// </summary>
public readonly record struct DeviceId(string Value)
{
    public override string ToString() => Value;

    public static DeviceId NewId() => new(Guid.NewGuid().ToString());
}

public readonly record struct WidgetId(string Value)
{
    public override string ToString() => Value;

    public static WidgetId NewId() => new(Guid.NewGuid().ToString());
}

public readonly record struct DashboardId(string Value)
{
    public override string ToString() => Value;

    public static DashboardId NewId() => new(Guid.NewGuid().ToString());
}

public readonly record struct GroupId(string Value)
{
    public override string ToString() => Value;

    public static GroupId NewId() => new(Guid.NewGuid().ToString());
}

File: android/domain/src/main/kotlin/com/controlldeck/domain/Ids.kt
package com.controlldeck.domain

/**
 * Strongly-typed identifiers. All are backed by plain Strings (UUIDs in
 * practice) but wrapped in inline value classes so the domain layer never
 * mixes up a DeviceId with a WidgetId etc. at compile time.
 */
@JvmInline
value class DeviceId(val value: String) {
    override fun toString(): String = value
}

@JvmInline
value class WidgetId(val value: String) {
    override fun toString(): String = value
}

@JvmInline
value class DashboardId(val value: String) {
    override fun toString(): String = value
}

@JvmInline
value class GroupId(val value: String) {
    override fun toString(): String = value
}

@JvmInline
value class AppId(val value: String) {
    override fun toString(): String = value
}
